### PR TITLE
#17305 Fix data sources config loading

### LINKED ISSUE #17279: Unable to Edit Connection v22.1.3
### Description

After upgrading from 22.0.0, I am unable to edit any existing connections. Selecting Edit Connection from the context menu (or using F4) does nothing.

There have been no changes to file permissions on the DBeaver directory (presently set at Full Control)

### DBeaver Version

Community Edition 22.1.3

### Operating System

Windows 10 21H2

### Database and driver

Oracle Database 11g Enterprise Edition Release 11.2.0.4.0 using Oracle JDBC driver v12.2.0.1.0

### Steps to reproduce

Highlight the desired connection, right-click and select Edit Connection - nothing happens.

### Additional context

_No response_

### LINKED ISSUE #17305: Failed to load existing database connections after update to 22.1.4.202208051447
### Description

After the Dbeaver is updated to 22.1.4, it failed to load the existing connections and shows the message `no default storage in registry General`

```
!ENTRY org.jkiss.dbeaver.model 4 0 2022-08-09 14:14:35.883
!MESSAGE Error loading datasource config from file:///Users/rveeranki/Library/DBeaverData/workspace6/General/.dbeaver/data-sources-default.json
!SUBENTRY 1 org.jkiss.dbeaver.model 4 0 2022-08-09 14:14:35.883
!MESSAGE no default storage in registry General (DataSourceRegistry)
!STACK 0
java.lang.IllegalStateException: no default storage in registry General (DataSourceRegistry)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.getDefaultStorage(DataSourceRegistry.java:173)
	at org.jkiss.dbeaver.registry.DataSourceSerializerModern.parseDataSources(DataSourceSerializerModern.java:529)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.loadDataSources(DataSourceRegistry.java:701)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.loadDataSources(DataSourceRegistry.java:657)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.loadDataSources(DataSourceRegistry.java:637)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.<init>(DataSourceRegistry.java:93)
	at org.jkiss.dbeaver.registry.DataSourceRegistry.<init>(DataSourceRegistry.java:86)
	at org.jkiss.dbeaver.registry.BaseProjectImpl.createDataSourceRegistry(BaseProjectImpl.java:179)
	at org.jkiss.dbeaver.registry.BaseProjectImpl.getDataSourceRegistry(BaseProjectImpl.java:171)
	at org.jkiss.dbeaver.model.navigator.DBNProject.readChildNodes(DBNProject.java:182)
	at org.jkiss.dbeaver.model.navigator.DBNResource.getChildren(DBNResource.java:197)
	at org.jkiss.dbeaver.model.navigator.DBNProject.getChildren(DBNProject.java:161)
	at org.jkiss.dbeaver.model.navigator.DBNProject.getDatabases(DBNProject.java:62)
	at org.jkiss.dbeaver.ui.navigator.database.DatabaseNavigatorView.getRootNode(DatabaseNavigatorView.java:87)
	at org.jkiss.dbeaver.ui.navigator.database.NavigatorViewBase.lambda$0(NavigatorViewBase.java:110)
	at org.jkiss.dbeaver.ui.UIExecutionQueue.executeInUI(UIExecutionQueue.java:52)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4345)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3968)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.jkiss.dbeaver.ui.app.standalone.DBeaverApplication.start(DBeaverApplication.java:281)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```

### DBeaver Version

22.1.4.202208051447

### Operating System

macOS Monterey 12.4

### Database and driver

1. Cloudera Impala cloudera_impala
2. Vertica vertica-jdbc-8

### Steps to reproduce

_No response_

### Additional context

_No response_

### LINKED ISSUE #17321: Loosing connections after upgrading from dbeaver < 6.1.3 to > 6.1.3
### Discussed in https://github.com/dbeaver/dbeaver/discussions/17304, https://github.com/dbeaver/dbeaver/issues/6599#issuecomment-1208920167

<div type='discussions-op-text'>

<sup>Originally posted by **blocka** August  9, 2022</sup>
When starting dbeaver the first time, it offered to migrate my configuration. This doesn't seem to have happened, as I don't see any connections. In the project directory `~/.local/share/DBeaverData/workspace6/General/` I see a .dbeaver-data-sources.xml file and an empty .project folder.</div>

```
java.version=11.0.12
java.vendor=Eclipse Foundation
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Command-line arguments:  -os linux -ws gtk -arch x86_64

org.jkiss.dbeaver.model
Error
Tue Aug 09 09:28:42 EDT 2022
Error loading datasource config from .local/share/DBeaverData/workspace6/General/.dbeaver-data-sources.xml
```

### PR DESCRIPTION
**Pull request by E1izabeth**
See https://github.com/dbeaver/dbeaver/wiki/Admin-Manage-Connections about data sources config location.

Also fixes https://github.com/dbeaver/dbeaver/issues/17321
Note: legacy data sources config should be located in project folder, modern data sources config is located in metadata folder (.dbeaver).
**_P.S. for testers: write me, I'll give you example of legacy config file._**
**_P.P.S. please, test team edition data sources loading also._**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceSerializerModern.java [hunk 1]
```diff
@@ -526,7 +527,7 @@ public void parseDataSources(
                     }
                     dataSource = new DataSourceDescriptor(
                         registry,
-                        registry.getDefaultStorage(),
+                        configurationStorage,
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceSerializerModern.java:531]: This change was intentional for the dbeaver/dbeaver-ee#1733. Please make sure this change doesn't break anything.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceSerializerModern.java:531]: We parse a data from `configurationStorage` here, not from some default storage, so, I'm sure that here should be `configurationStorage` which data we are reading.
The `getConfigurationStorages()` method is used both to get a collection of `configurationStorage`, which data we parse here (for each of them separately in this method), and in `registry.getDefaultStorage()`, which was previously used, with the only difference - `registry.getDefaultStorage()` chooses the default one from all configuration storages. In the same time `DataSourceRemoteStorage::isDefault` method always returns `true`, so all remote storages (if if there could be more than one for cloud) are default.

Of course, it's better to test both ce/ee and team edition. I added a note for testers,

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java [hunk 2]
```diff
@@ -110,11 +110,15 @@ public List<DBPDataSourceConfigurationStorage> getConfigurationStorages() {
 
     @Override
     public InputStream readConfiguration(@NotNull String name) throws IOException {
-        Path path = getConfigurationPath(false).resolve(name);
-        if (!Files.exists(path)) {
+        Path configPath = getConfigurationPath(false).resolve(name);
+        if (!Files.exists(configPath)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java]:
```suggestion
        if (Files.notExists(configPath)) {
```

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java [hunk 3]
```diff
@@ -110,11 +110,15 @@ public List<DBPDataSourceConfigurationStorage> getConfigurationStorages() {
...
+            // maybe it's .dbeaver-data-sources*.xml in the project folder (DBeaver < 6.1.3 (Legacy))
+            configPath = project.getAbsolutePath().resolve(name);
+        }
+        if (!Files.exists(configPath)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java]:
```suggestion
        if (Files.notExists(configPath)) {
```

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java [hunk 4]
```diff
@@ -110,11 +110,15 @@ public List<DBPDataSourceConfigurationStorage> getConfigurationStorages() {
 
     @Override
     public InputStream readConfiguration(@NotNull String name) throws IOException {
-        Path path = getConfigurationPath(false).resolve(name);
-        if (!Files.exists(path)) {
+        Path configPath = getConfigurationPath(false).resolve(name);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceConfigurationManagerNIO.java]: `configPath` -> `path` to reduce code changes

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Comment by uslss** [comment]
still reproducible

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java [hunk 5]
```diff
@@ -170,7 +173,16 @@ DBPDataSourceConfigurationStorage getDefaultStorage() {
...
-            throw new IllegalStateException("no default storage in registry " + this);
+            log.warn("no default storage in registry " + this);
+            try {
+                java.nio.file.Path configPath = this.getProject().getMetadataFolder(false).resolve(MODERN_CONFIG_FILE_NAME);
+                Files.createFile(configPath);
+                DBPDataSourceConfigurationStorage defaultStorage = new DataSourceFileStorage(configPath, false, true);
+                this.storages.add(defaultStorage);
+                return defaultStorage;
+            } catch (IOException e) {
+                throw new IllegalStateException("no default storage in registry " + this + "and unable to create it");
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java]:
```suggestion
                throw new IllegalStateException("Unable to create a default storage in registry " + this, e);
```

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java [hunk 6]
```diff
@@ -170,7 +173,16 @@ DBPDataSourceConfigurationStorage getDefaultStorage() {
                 }
             }
             // No default storage. Seems to be an internal error
-            throw new IllegalStateException("no default storage in registry " + this);
+            log.warn("no default storage in registry " + this);
+            try {
+                java.nio.file.Path configPath = this.getProject().getMetadataFolder(false).resolve(MODERN_CONFIG_FILE_NAME);
+                Files.createFile(configPath);
+                DBPDataSourceConfigurationStorage defaultStorage = new DataSourceFileStorage(configPath, false, true);
+                this.storages.add(defaultStorage);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java:181]: Looks fragile and hack-ish. I'm not sure 😐

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java:181]: We already add default storage in `DataSourceConfigurationManagerNIO::getConfigurationStorages()` if there is no one. I guess someone just forgot to create a new file for this case.
```
if (storages.isEmpty()) {
    storages.add(
        new DataSourceFileStorage(
            metadataFolder.resolve(DataSourceRegistry.MODERN_CONFIG_FILE_NAME), false, true));
}
```
Anyway I see default storage everywhere. If I correctly understand the code, it wasn't assumed that we could not have default storage or this case wasn't correctly handled.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceRegistry.java:181]: Maybe I could handle it there.

**Comment by uslss** [comment]
verified

**Pull request merged**